### PR TITLE
[https://nvbugs/6625851][fix] Fail guided-decoding requests reaching a dead-end grammar state

### PR DESCRIPTION
**Pull request by zhaoyangwang-nvidia**
## Description

Under sustained concurrent grammar-constrained load with `guided_decoding_backend: xgrammar`, a single request can hard-kill an entire deployment (nvbugs/6625851: 16-rank DEP16 job, `MPI_Abort` errorcode 137).

When a grammar matcher reaches a state with no valid next token, `fill_next_token_bitmask` produces an all-zero row. `logits_bitmask` then masks that entire logits row to `-inf`, `softmax` returns NaN for it, and the sampler's device-side NaN assert (`_flashinfer_check_nans` in `sampler_strategy.py`) fires. That assert is a *global* device assert, so one request's grammar dead-end escalates into a peer-kill of every rank. Because `_assert_async` is asynchronous, the reported stack sits downstream of the faulting kernel, which is why the crash surfaces at `sampler.update_requests` -> `sampler_event.synchronize()` with no apparent link to guided decoding.

This PR detects the all-zero row on the host, immediately after `fill_next_token_bitmask`, and fails that one request through the guided decoder's existing `failed_requests` error path.

### Why detect at fill time on the host

vLLM and vSGLang both treat a stuck grammar matcher as a per-request failure and never let it reach a fatal assert. vLLM sets `RequestStatus.FINISHED_ERROR` for that request only when `grammar.accept_tokens()` returns False (`v1/core/sched/scheduler.py`); SGLang sets `FINISH_ABORT()` for that request only when `accept_token` raises (`managers/scheduler_components/batch_result_processor.py`), and its NaN async assert is env-gated and off in production. Neither scans the bitmask for emptiness, on host or device.

Detecting at fill time keeps the check where the row is already being written, so there is no GPU work and no added kernel launches on the decode critical path. It also means no grammar-violating token is ever emitted: a row with no valid continuation cannot produce valid output, so the request fails explicitly instead of silently returning a token that breaks the schema.

Only bits below `vocab_size_padded` are counted. The trailing bits of a partial last word are never read by the apply kernel and are not guaranteed to be cleared by the backends, so counting them would report a dead-end row as valid.

For draft requests the dead end terminates drafting instead of failing the request, consistent with the existing unacceptable-draft-token handling. The matcher has already advanced past `new_token` on that path, so the advance is recorded in `num_advanced_draft_tokens` and `_rollback_draft_tokens` still undoes it.

### Scope and follow-ups

This PR fixes the producer of the NaN and deliberately does not touch the sampler. The underlying hazard remains: `_flashinfer_check_nans` issues an unconditional global device assert, so any other NaN source (fp16 activation overflow, custom logit processors, draft logits) can still take down a whole deployment. Gating that assert behind an env var, as SGLang does, is left to a separate PR.

Re-enabling the Kimi K3 strict-tool grammar by default (`TRTLLM_KIMI_K3_STRICT_TOOL_GRAMMAR`, added in #17845 as a mitigation for this same bug) is also out of scope here: it needs validation on the original 16xGB300 workload first.

## Test Coverage

`tests/unittest/_torch/misc/test_guided_decoder_bitmask.py` (22 tests, all passing locally on B200).

Unit coverage of `row_has_valid_token`:
- `test_dead_end_row_has_no_valid_token` -- an all-zero row is reported dead, for word-aligned and partial-last-word vocab sizes
- `test_single_valid_token_is_detected` -- a single set bit anywhere below `vocab_size_padded` keeps the row alive
- `test_trailing_padding_bits_do_not_count` -- padding bits above `vocab_size_padded` do not mark a dead row valid, while the highest in-range bit of that same partial word still does
- `test_sign_bit_counts_as_valid_token` -- bit 31, which makes the `int32` word negative, counts as a valid token

Behavior coverage driving the real `_build` with a scripted fake matcher:
- `test_build_fails_request_on_dead_end_row` -- the request lands in `failed_requests`, its row stays unguided, and `num_advanced_tokens` is cleared so `_rollback_rejected_tokens` skips it
- `test_build_draft_dead_end_is_rolled_back` -- drafting terminates and `_rollback_draft_tokens` actually issues `rollback(1)`
- `test_build_draft_position_dead_end_stops_guiding` -- a dead end at a later draft position leaves the remaining positions unguided without failing the request
- `test_build_dead_end_isolates_the_failing_request` -- in a two-request batch only the dead-end request fails and the healthy request keeps its guided row

Verified non-vacuous by mutation: removing the `num_advanced_draft_tokens` accumulation fails exactly `test_build_draft_dead_end_is_rolled_back`, and removing the dead-end check entirely fails all four `_build` tests and no others.

Not covered: the tests use a scripted fake matcher rather than a real xgrammar matcher, and the fix has not been reproduced against the original 16xGB300 DEP16 workload.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 1]
```diff
@@ -257,6 +281,20 @@ def _build(self, requests: GuidedRequests) -> List[Tuple[int, str]]:
                 self.num_advanced_tokens[slot] += 1
                 if not matcher.is_terminated():
                     matcher.fill_next_token_bitmask(self.bitmask_host, offset)
+                    if not self._has_valid_token(offset):
+                        if req.is_draft:
+                            self.is_draft_terminated[slot] = True
+                            logger.debug(
+                                f"Draft request {req.request_id} at slot {slot} reached a grammar state with no valid token."
+                            )
+                            continue
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:295]: Confirmed and fixed in 1fa54be. The matcher had already advanced past `new_token` on this path, unlike the unacceptable-token `continue` just above it where `accept_token` returned False and the counter is still 0. I copied that structure without accounting for the differing precondition, so `_rollback_draft_tokens` under-rolled by one token and left the target model a token ahead. `test_build_draft_dead_end_is_rolled_back` now pins this: removing the accumulation again fails that test and only that test.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 2]
```diff
@@ -257,6 +281,20 @@ def _build(self, requests: GuidedRequests) -> List[Tuple[int, str]]:
...
+                        # The request is about to be terminated, so exclude it
+                        # from the rollback pass: the matcher advance made
+                        # above is never verified against accepted tokens.
+                        self.num_advanced_tokens[slot] = 0
+                        raise ValueError(
+                            f"Request {req.request_id} at slot {slot} reached a grammar state with no valid token."
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:301]: Added in 1fa54be. `tests/unittest/_torch/misc/test_guided_decoder_bitmask.py` now drives `_build` directly with a scripted fake matcher: `test_build_fails_request_on_dead_end_row` asserts the request lands in `failed_requests`, the row stays unguided, and `num_advanced_tokens` is cleared; `test_build_draft_dead_end_is_rolled_back` asserts the draft path terminates drafting and that `_rollback_draft_tokens` actually issues `rollback(1)`.

File: tensorrt_llm/_torch/pyexecutor/guided_decoder.py [hunk 3]
```diff
@@ -273,6 +311,11 @@ def _build(self, requests: GuidedRequests) -> List[Tuple[int, str]]:
                             break
                         matcher.fill_next_token_bitmask(self.bitmask_host,
                                                         offset + i)
+                        if not self._has_valid_token(offset + i):
+                            # Stop guiding here rather than failing the
+                            # request: the remaining draft positions are left
+                            # unconstrained and get verified as usual.
+                            break
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tensorrt_llm/_torch/pyexecutor/guided_decoder.py:323]: Added in 1fa54be as `test_build_draft_position_dead_end_stops_guiding`: one draft token is accepted, the next row is empty, and the test asserts the request does not fail, only position 0 is guided, and the remaining draft positions keep `token_mask` at 0. 8c74b34 additionally makes the fake matcher return a valid row past the end of its script, so that removing the check fails this test on its assertions rather than on the helper running out of scripted rows.

File: tests/unittest/_torch/misc/test_guided_decoder_bitmask.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    requests = GuidedRequests(
+        [_generation_request(is_draft=False, draft_tokens=[])],
+        num_contexts=0,
+        num_generations=1,
+        max_num_draft_tokens=0,
+    )
+
+    failed_requests = decoder._build(requests)
+
+    assert [req_id for req_id, _ in failed_requests] == [7]
```
[the remark this answers is not in the document]
Comment (zhaoyangwang-nvidia) [reply tests/unittest/_torch/misc/test_guided_decoder_bitmask.py:169]: Good catch - this is the property the fix exists for, and a single-request batch could not distinguish it. Added `test_build_dead_end_isolates_the_failing_request` in 448316f: two requests, the dead-end one in slot 0 and a healthy one in slot 1, asserting `failed_requests == [7]` and that slot 1 keeps its guided row (`token_mask_host[1] == 1`, `num_guided_tokens[1] == 1`).

**Review by BowenFu — APPROVED** [review]

**Review by tongyuantongyu — APPROVED** [review]
